### ISSUE TITLE
EGM pose controller

### ISSUE THREAD

**Issue by cpilon** [ros-industrial/abb_robot_driver#10]
I was wondering if there was any example available for a EGM pose controller?

I have been able to follow example 2 and use the _joint_group_velocity_controller_ and _joint_group_position_controller_. However I called the _/egm/controller_manager/list_controller_types_ service and these does not seem to be a pose controller in the list.

**Comment by gavanderhoorn** [comment]
If by "pose controller" you mean a Cartesian controller, then you are right, there is no example for those.

But that's more because there is no standard/default Cartesian controller in `ros_controllers` than anything else.

Controllers which require special "Cartesian resources" will not be supported out-of-the-box, but you could perhaps see whether [fzi-forschungszentrum-informatik/cartesian_ros_control](https://github.com/fzi-forschungszentrum-informatik/cartesian_ros_control) covers your use-case.

Finally, just to clarify: `abb_robot_driver` is a `ros_control` based ROS driver for ABB robots. It's not an EGM application or library, nor necessarily an EGM-only ROS component.

> However I called the `/egm/controller_manager/list_controller_types` service and these does not seem to be a pose controller in the list.

The output of that service is the list of controllers configured in the `ros_control` configuration loaded by the respective `.launch` files in the bringup package. It does not list whatever control interfaces EGM supports.

**Comment by cpilon** [comment]
@gavanderhoorn - Thanks for the clarification on the `ros_controllers`.
It seems the best approach for my application is to use `abb_libegm` directly.

**Comment by gavanderhoorn** [comment]
Adding a Cartesian resource and then using a forwarding Twist controller on-top of that for instance should really not be that much work.

Probably less work than using EGM directly.

**Comment by gavanderhoorn** [comment]
And I forgot to mention [fzi-forschungszentrum-informatik/cartesian_controllers](https://github.com/fzi-forschungszentrum-informatik/cartesian_controllers).

Those should work with the joint space interfaces already supported by `abb_robot_driver`.

**Comment by cpilon** [comment]
I already had a working code based on `abb_libegm` before starting using `abb_robot_driver` so in my case it was easy but I will definitely consider your recommendation as an improvement.

**Comment by gavanderhoorn** [comment]
If you do, we'd be interested in potentially integrating something like that here.

Adding Cartesian support is on the TODO list, so you'd help a lot of other users.

**Comment by cpilon** [comment]
Of course! I will share if I come up with anything useful.

**Comment by gavanderhoorn** [comment]
As it seems we've answered the initial question, I'm going to close the issue.

Feel free to keep commenting on it of course.

**Comment by DyumanAditya** [comment]
Hi, from what I understood here, if I have a cartesian pose that I want my robot end-effector to move to, I cannot send this pose via the abb_robot_driver to RobotStudio? I would have to use [another package](https://github.com/fzi-forschungszentrum-informatik/cartesian_controllers) with the abb_robot_driver to get this to work?

**Comment by gavanderhoorn** [comment]
If you're not interested in getting to that pose while actively trying to avoid collisions -- so just use pure Cartesian interpolation from start to end -- then, yes, you'll have to configure your driver to load a Cartesian controller.

`abb_robot_driver` currently does not support Cartesian resources, as until *very* recently, there was no support for that in `ros_control`.

**Comment by DyumanAditya** [comment]
>If you're not interested in getting to that pose while actively trying to avoid collisions
Did you mean "If I AM interested" ?

What exactly does the service `/rws/sm_addin/start_egm_pose` do ? And how can I use it?

If I wanted to let RobotStudio handle the Inverse Kinematics and somehow send it a pose, how can I achieve this using this package? Would the service `/rws/sm_addin/start_egm_pose` help?

**Comment by gavanderhoorn** [comment]
> > If you're not interested in getting to that pose while actively trying to avoid collisions
> 
> Did you mean "If I AM interested" ?

No, I meant to ask what I wrote.

The IRC5 will use straightforward linear interpolation between poses. There is no collision detection nor avoidance, except for the robot with itself and -- depending on whether you have that option -- what SafeMove can do for you.

So using EGM in pose mode will also just linearly interpolate between poses.

This in contrast to something like MoveIt, which will try to plan a collision free path for you between your starting and destination pose. MoveIt can directly use the driver as-is, as it uses `JointTrajectory` goals, which are fully supported.

> What exactly does the service `/rws/sm_addin/start_egm_pose` do ?

It starts an EGM motion which accepts Cartesian poses as targets.

> And how can I use it?

By sending Cartesian poses to the controller.

But not with `abb_robot_driver`, because, as I wrote before:

> `abb_robot_driver` currently does not support Cartesian resources, as until _very_ recently, there was no support for that in `ros_control`.

that comment probably also answers your next question:

> If I wanted to let RobotStudio handle the Inverse Kinematics and somehow send it a pose, how can I achieve this using this package? Would the service `/rws/sm_addin/start_egm_pose` help?

As some general advice: if you aren't already, it would be good to familiarise yourself with the basic workings of EGM.

EGM is not something we invented here, or for this driver, but a regular software option developed and supported by ABB Robotics. We merely use it to "talk" to the IRC5.

Most of the functionality offered by the driver is a thin wrapper around EGM and RWS. If you don't have any experience with either of these two technologies, nor with ABB robots in general, making sense of the services, topics and terminology is going to be difficult.

`<my_personal_opinion>`

> This driver (ie: `abb_robot_driver`) should not be your first experience with ABB robots. Industrial robots can be dangerous and their correct functioning depends on how you use them. Additionally, there are many aspects in the configuration which will influence the way they behave, which, when adding ROS to the mix, can make things quite complex.
> 
> If you have the option, I would recommend to first get your bearings with plain RobotStudio and the operation of ABB robots, before adding ROS and this driver. Most likely you'll find many things much clearer and easier to use.

`</my_personal_opinion>`

**Comment by DyumanAditya** [comment]
Thanks for the detailed response. 
I need some advice as to how I can proceed: As it turns out, I need straightforward linear interpolation between poses _without_ collision detection and other planning (the most rudimentary form of cartesian control). What is the best way to accomplish this using ROS and RobotStudio (without using rapid)? 
So far I've understood from what you've said that this is not possible using the `abb_robot_driver`?
But this would be simpler if I were to use `abb_libegm` ?

**Comment by gavanderhoorn** [comment]
Didn't you already find [fzi-forschungszentrum-informatik/cartesian_controllers](https://github.com/fzi-forschungszentrum-informatik/cartesian_controllers)?

That's a collection of `ros_control` controllers which accept Cartesian inputs.

> So far I've understood from what you've said that this is not possible using the `abb_robot_driver`?

`ros_control` doesn't / didn't support Cartesian interfaces, so `abb_robot_driver` doesn't either.

[fzi-forschungszentrum-informatik/cartesian_ros_control](https://github.com/fzi-forschungszentrum-informatik/cartesian_ros_control) adds Cartesian support to `ros_control`, but that would need to be added to `abb_robot_driver` to support it.

> But this would be simpler if I were to use `abb_libegm` ?

Well, that would no longer be "using ROS". `abb_libegm` is a stand-alone library for interfacing with EGM capable ABB robot controllers. Nothing more.

**Comment by gavanderhoorn** [comment]
If you have any further questions about this, please open a new issue. We should try to avoid commenting on an already closed issue.
